assets/quick-order-list.js

```diff
@@ -1,307 +1,281 @@
-          const variantIds = [];
-          this.querySelectorAll('.variant-item').forEach((item) => {
-            variantIds.push(parseInt(item.dataset.variantId));
-          });
-          if (
-            event.source === this.quickOrderListId ||
-            !event.cartData.items?.some((element) => variantIds.includes(element.variant_id))
-          ) {
-            return;
-          }
-          // If its another section that made the update
-          this.refresh().then(() => {
-            this.defineInputsAndQuickOrderTable();
-            this.addMultipleDebounce();
-          });
+        this.cartUpdateUnsubscriber = subscribe(PUB_SUB_EVENTS.cartUpdate, async (event) => {
+          // skip if cart event was triggered by this section
+          if (event.source === this.id) return;
+
+          this.toggleTableLoading(true);
+          await this.refresh();
+          this.toggleTableLoading(false);
         });
-        this.sectionId = this.dataset.section;
+
+        this.initEventListeners();
       }
 
       disconnectedCallback() {
         this.cartUpdateUnsubscriber?.();
       }
 
-      defineInputsAndQuickOrderTable() {
+      initEventListeners() {
+        this.querySelectorAll('.pagination__item').forEach((link) => {
+          link.addEventListener('click', async (event) => {
+            event.preventDefault();
+            event.stopPropagation();
+
+            const url = new URL(event.currentTarget.href);
+
+            this.toggleTableLoading(true);
+            await this.refresh(url.searchParams.get('page') || '1');
+            this.scrollTop();
+            this.toggleTableLoading(false);
+          });
+        });
+
+        this.querySelector('.quick-order-list__table').addEventListener('focusin', this.switchVariants.bind(this));
+
+        this.initVariantEventListeners();
+      }
+
+      initVariantEventListeners() {
         this.allInputsArray = Array.from(this.querySelectorAll('input[type="number"]'));
-        this.quickOrderListTable = this.querySelector('.quick-order-list__table');
-        this.quickOrderListTable.addEventListener('focusin', this.switchVariants.bind(this));
+
+        this.querySelectorAll('quantity-input').forEach((qty) => {
+          const debouncedOnChange = debounce(this.onChange.bind(this), 250);
+          qty.addEventListener('change', debouncedOnChange);
+        });
+
+        this.querySelectorAll('.quick-order-list-remove-button').forEach((button) => {
+          button.addEventListener('click', (event) => {
+            event.preventDefault();
+            this.toggleLoading(true);
+            this.startQueue(button.dataset.index, 0);
+          });
+        });
+      }
+
+      get currentPage() {
+        return this.querySelector('.pagination-wrapper')?.dataset?.page ?? '1';
+      }
+
+      get cartVariantsForProduct() {
+        return JSON.parse(this.querySelector('[data-cart-contents]')?.innerHTML || '[]');
```

Comment:
This is the secret sauce for remove all. Where we used to read variant inputs with non-zero quantities out of the DOM (which doesn't work with paginated variants), we now grab the JSON dump of this cart product ([liquid change](https://github.com/Shopify/dawn/pull/3710/files#diff-544215e8ade6cd5533a29ec01cff17d5dd80698302f9fdc9828b774306f8a198R288-R290)).